lib/rack/utils.rb

```diff
@@ -248,71 +248,59 @@ def add_cookie_to_header(header, key, value)
       end
       value = [value] unless Array === value
 
-      cookie = "#{escape(key)}=#{value.map { |v| escape v }.join('&')}#{domain}" \
+      return "#{escape(key)}=#{value.map { |v| escape v }.join('&')}#{domain}" \
         "#{path}#{max_age}#{expires}#{secure}#{httponly}#{same_site}"
+    end
 
-      case header
-      when nil, ''
-        cookie
-      when String
-        [header, cookie].join("\n")
-      when Array
-        (header + [cookie]).join("\n")
+    def set_cookie_header!(headers, key, value)
+      if header = headers[SET_COOKIE]
+        if header.is_a?(Array)
+          header << set_cookie_header(key, value)
+        else
+          headers[SET_COOKIE] = [header, set_cookie_header(key, value)]
+        end
       else
-        raise ArgumentError, "Unrecognized cookie header value. Expected String, Array, or nil, got #{header.inspect}"
+        headers[SET_COOKIE] = set_cookie_header(key, value)
       end
     end
 
-    def set_cookie_header!(header, key, value)
-      header[SET_COOKIE] = add_cookie_to_header(header[SET_COOKIE], key, value)
-      nil
-    end
-
-    def make_delete_cookie_header(header, key, value)
-      case header
-      when nil, ''
-        cookies = []
-      when String
-        cookies = header.split("\n")
-      when Array
-        cookies = header
-      end
-
-      key = escape(key)
-      domain = value[:domain]
-      path = value[:path]
-      regexp = if domain
-                 if path
-                   /\A#{key}=.*(?:domain=#{domain}(?:;|$).*path=#{path}(?:;|$)|path=#{path}(?:;|$).*domain=#{domain}(?:;|$))/
+    # Adds a cookie that will *remove* a cookie from the client.  Hence the
+    # strange method name.
+    def delete_set_cookie_header(key, value = {})
+      set_cookie_header(key, {
+        value: '', path: nil, domain: nil,
+        max_age: '0',
+        expires: Time.at(0)
+      }.merge(value))
+    end
+
+    def delete_set_cookie_header!(header, key, value = {})
+      if header
+        header = Array(header)
+
+        key = escape(key)
+        domain = value[:domain]
+        path = value[:path]
+        regexp = if domain
+                   if path
+                     /\A#{key}=.*(?:domain=#{domain}(?:;|$).*path=#{path}(?:;|$)|path=#{path}(?:;|$).*domain=#{domain}(?:;|$))/
+                   else
+                     /\A#{key}=.*domain=#{domain}(?:;|$)/
+                   end
+                 elsif path
+                   /\A#{key}=.*path=#{path}(?:;|$)/
                  else
-                   /\A#{key}=.*domain=#{domain}(?:;|$)/
+                   /\A#{key}=/
```

Comment:
This could all go if we don't care about deleting cookies from the `set-header` header. Unless there is some RFC that specifically calls out this behaviour, I don't think we should do it like this.